ansible/roles/gpu/tasks/nvidia-gpu.yaml

```diff
@@ -48,6 +48,58 @@
     path: "{{ nvidia_remote_bundle_path }}"
     state: directory
 
+- name: find nvidia driver installer file
+  find:
+    paths: "{{ nvidia_remote_bundle_path }}"
+    patterns: "NVIDIA-Linux-x86_64-*.run"
+  register: nvidia_driver_files
+
+- name: extract version from file names
+  set_fact:
+    nvidia_driver_versions: "{{ nvidia_driver_files.files | map(attribute='path') | map('regex_replace', '^.*NVIDIA-Linux-x86_64-(\\d+\\.\\d+\\.\\d+)\\.run$', '\\1') | list }}"
+    uninstall: false
+
+- name: compare versions with nvidia_driver_version
+  loop: "{{ nvidia_driver_versions }}"
+  when: item is version( nvidia_driver_version , '<')
+  set_fact:
+    uninstall: true
+    previous_version: "{{ item }}"
+
+- name: uninstall and stop programs using older nvidia drivers
+  block:
+  - name: uninstall older versions
+    shell: "{{ nvidia_remote_bundle_path }}/NVIDIA-Linux-x86_64-{{ previous_version }}.run --uninstall -s"
+    args:
+      chdir: "{{ nvidia_remote_bundle_path }}"
+    become: yes
+
+  # this is to make installing new drivers easier. older kernels don't support unloading modules in use
+  - name: stop persistenced service before installing the new drivers
+    ansible.builtin.systemd:
+      daemon_reload: yes
+      name: nvidia-persistenced
+      state: stopped
+    become: yes
+
+  - name: uninstall older versions
```

Comment:
This task (`uninstall older versions`) is executed again after stopping the service. is that intentional?

Reply:
fixed